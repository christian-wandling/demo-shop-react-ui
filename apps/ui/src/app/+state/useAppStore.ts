import { create } from 'zustand';
import { useUserStore } from '@demo-shop-react-ui/user';
import { initKeycloak } from '@demo-shop-react-ui/auth';
import { authConfig } from '../config/config';
import { useShoppingCartStore } from '@demo-shop-react-ui/shopping';

interface AppState {
  isInitialized: boolean;
  initialize: () => Promise<void>;
}

export const useAppStore = create<AppState>()((set, get) => {
  return {
    isInitialized: false,

    initialize: async () => {
      const { fetchCurrentUser } = useUserStore.getState();
      const { fetchCurrentSession } = useShoppingCartStore.getState();
      const hasActiveSession = await initKeycloak(authConfig);

      if (hasActiveSession) {
        await fetchCurrentUser();
        await fetchCurrentSession();
      }

      set({ isInitialized: true });
    },
  };
});
