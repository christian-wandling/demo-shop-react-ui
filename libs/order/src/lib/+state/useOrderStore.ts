import { create } from 'zustand';
import { getApi, OrderResponse, OrderStatus } from '@demo-shop-react-ui/api';

interface OrderState {
  orders: OrderResponse[];

  getSortedOrders: () => OrderResponse[];
  getOrderById: (id: number) => OrderResponse | undefined;
  fetchOrders: () => Promise<void>;
  fetchOrderById: (id: number) => Promise<void>;
}

export const useOrderStore = create<OrderState>()((set, get) => {
  return {
    orders: [],

    fetchOrders: async () => {
      const { orderApi } = getApi();
      const res = await orderApi.getAllOrdersOfCurrentUser();
      set({ orders: res.items });
    },

    fetchOrderById: async (id: number) => {
      const { orderApi } = getApi();
      const order = await orderApi.getOrderById({ id });
      set(state => ({ orders: [...state.orders, order] }));
    },

    getSortedOrders: () => {
      const { orders } = get();

      return orders.sort((a, b) => {
        if (a.status === b.status) {
          return a.created > b.created ? -1 : 1;
        }

        return a.status === OrderStatus.Created ? -1 : 1;
      });
    },

    getOrderById: id => {
      return get().orders.find(order => order.id === id);
    },
  };
});
