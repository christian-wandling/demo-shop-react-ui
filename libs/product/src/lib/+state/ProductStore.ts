import { create } from 'zustand';
import { Configuration, ProductApi, ProductResponse } from '@demo-shop-react-ui/api';
import { AllowedProductFilterTypes, ProductFilter } from '../models/product-filter';

interface ProductState {
  products: ProductResponse[];
  filter: ProductFilter;

  setFilter: (filter: ProductFilter) => void;
  getFilteredProducts: () => ProductResponse[];
  getProductById: (id: number) => ProductResponse | undefined;
  fetchProducts: () => Promise<void>;
  fetchProductById: (id: number) => Promise<void>;
}

const api = new ProductApi(new Configuration({ basePath: '' }));

export const useProductStore = create<ProductState>()((set, get) => ({
  products: [],
  filter: {},

  fetchProducts: async () => {
    const res = await api.getAllProducts();
    set({ products: res.items });
  },

  fetchProductById: async (id: number) => {
    const product = await api.getProductById({ id });
    set(state => ({ products: [...state.products, product] }));
  },

  getFilteredProducts: () => {
    const { filter, products } = get();
    const filterValues = Object.entries(filter);

    if (filterValues.length === 0) {
      return products;
    }

    return products.filter(entity =>
      filterValues.every(([key, filterValue]) => {
        const value = entity[key as AllowedProductFilterTypes];

        if (Array.isArray(value)) {
          return value.some(item => item.toLowerCase().includes(filterValue.toLowerCase()));
        }

        return value.toLowerCase().includes(filterValue.toLowerCase());
      })
    );
  },

  getProductById: id => {
    return get().products.find(product => product.id === id);
  },

  setFilter: (filter: ProductFilter) => {
    set(state => ({ ...state, filter }));
  },
}));
