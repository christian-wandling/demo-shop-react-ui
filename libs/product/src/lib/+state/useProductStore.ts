import { create } from 'zustand';
import { getApi, ProductResponse } from '@demo-shop-react-ui/api';
import { AllowedProductFilterTypes, ProductFilter } from '../models/productFilter';

interface ProductState {
  products: ProductResponse[];
  filter: ProductFilter;

  setFilter: (filter: ProductFilter) => void;
  getFilteredProducts: () => ProductResponse[];
  getProductById: (id: number) => ProductResponse | undefined;
  fetchProducts: () => Promise<void>;
  fetchProductById: (id: number) => Promise<void>;
}

export const useProductStore = create<ProductState>()((set, get) => {
  return {
    products: [],
    filter: {
      name: '',
    },

    fetchProducts: async () => {
      const { productApi } = getApi();
      const res = await productApi.getAllProducts();
      set({ products: res.items });
    },

    fetchProductById: async (id: number) => {
      const { productApi } = getApi();
      const product = await productApi.getProductById({ id });
      set(state => ({ products: [...state.products, product] }));
    },

    getFilteredProducts: () => {
      const { filter, products } = get();
      const filterValues = Object.entries(filter);

      if (filterValues.length === 0) {
        return products;
      }

      return products.filter(entity =>
        filterValues.every(([key, filterValue]) => {
          const value = entity[key as AllowedProductFilterTypes];

          if (Array.isArray(value)) {
            return value.some(item => item.toLowerCase().includes(filterValue.toLowerCase()));
          }

          return value.toLowerCase().includes(filterValue.toLowerCase());
        })
      );
    },

    getProductById: id => {
      return get().products.find(product => product.id === id);
    },

    setFilter: (filter: ProductFilter) => {
      set(state => ({ ...state, filter }));
    },
  };
});
