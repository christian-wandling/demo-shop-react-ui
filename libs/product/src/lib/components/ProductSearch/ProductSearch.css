.search-container:not(:hover) .search-input:not(.extended) input {
  width: 0;
  padding-left: 0;
  padding-right: 0;
}
