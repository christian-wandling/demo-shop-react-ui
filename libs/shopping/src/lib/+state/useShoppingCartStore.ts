import { create } from 'zustand';
import { AddCartItemRequest, CartItemResponse, getApi, UpdateCartItemQuantityRequest } from '@demo-shop-react-ui/api';

interface ShoppingCartState {
  items: CartItemResponse[];
  sessionId: number | null;
  showCart: boolean;
  fetchCurrentSession: () => Promise<void>;
  add: (addCartItemRequest: AddCartItemRequest) => Promise<void>;
  remove: (id: number) => Promise<void>;
  updateItemQuantity: (id: number, updateCartItemQuantityRequest: UpdateCartItemQuantityRequest) => Promise<void>;
  getItemById: (id: number) => CartItemResponse | undefined;
  getItemByProductId: (productId: number) => CartItemResponse | undefined;
  setShowCart: (showCart: boolean) => void;

  getTotalPrice: () => number;

  getItemCount: () => number;

  hasActiveSession: () => boolean;
}

export const useShoppingCartStore = create<ShoppingCartState>()((set, get) => {
  return {
    items: [],
    sessionId: null,
    showCart: false,

    getTotalPrice(): number {
      return get().items.reduce((sum, item) => sum + item.totalPrice, 0);
    },

    getItemCount(): number {
      return get().items.reduce((sum, item) => sum + item.quantity, 0);
    },

    hasActiveSession(): boolean {
      return get().sessionId !== null;
    },

    fetchCurrentSession: async () => {
      const { shoppingSessionApi } = getApi();
      const session = await shoppingSessionApi.resolveCurrentShoppingSession();
      set({ sessionId: session.id, items: session.items });
    },

    add: async (addCartItemRequest: AddCartItemRequest) => {
      if (!get().hasActiveSession()) {
        throw new Error('No active shopping session');
      }

      const itemInCart = get().getItemByProductId(addCartItemRequest.productId);

      if (itemInCart) {
        get().updateItemQuantity(itemInCart.id, { quantity: itemInCart.quantity + 1 });
        return;
      }

      const { shoppingSessionApi } = getApi();
      await shoppingSessionApi.addCartItem({ addCartItemRequest });
      get().fetchCurrentSession();
    },

    remove: async (id: number) => {
      if (!get().hasActiveSession()) {
        throw new Error('No active shopping session');
      }

      const itemInCart = get().getItemById(id);

      if (!itemInCart) {
        throw new Error('Item not found');
      }

      if (itemInCart.quantity > 1) {
        get().updateItemQuantity(itemInCart.id, { quantity: itemInCart.quantity - 1 });
        return;
      }

      const { shoppingSessionApi } = getApi();
      await shoppingSessionApi.removeCartItem({ id });
      get().fetchCurrentSession();
    },

    updateItemQuantity: async (id: number, updateCartItemQuantityRequest: UpdateCartItemQuantityRequest) => {
      if (!get().hasActiveSession()) {
        throw new Error('No active shopping session');
      }

      const { shoppingSessionApi } = getApi();
      await shoppingSessionApi.updateCartItemQuantity({ id, updateCartItemQuantityRequest });
      get().fetchCurrentSession();
    },

    getItemById: id => {
      return get().items.find(item => item.id === id);
    },

    getItemByProductId: productId => {
      return get().items.find(item => item.productId === productId);
    },

    setShowCart: (showCart: boolean) => {
      set(state => ({ ...state, showCart }));
    },
  };
});
