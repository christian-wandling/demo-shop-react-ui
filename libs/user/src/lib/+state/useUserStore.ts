import { create } from 'zustand';
import { getApi, UpdateUserAddressRequest, UpdateUserPhoneRequest, UserResponse } from '@demo-shop-react-ui/api';

interface UserState {
  user: UserResponse | null;

  fetchCurrentUser: () => Promise<void>;
  updateUserAddress: (address: UpdateUserAddressRequest) => Promise<void>;
  updateUserPhone: (address: UpdateUserPhoneRequest) => Promise<void>;
}

export const useUserStore = create<UserState>()((set, get) => {
  return {
    user: null,

    fetchCurrentUser: async () => {
      const { userApi } = getApi();
      const user = await userApi.resolveCurrentUser();
      set({ user });
    },

    updateUserAddress: async (updateUserAddressRequest: UpdateUserAddressRequest) => {
      const { userApi } = getApi();
      const address = await userApi.updateCurrentUserAddress({ updateUserAddressRequest });
      set(state => ({ user: state.user && { ...state.user, address } }));
    },

    updateUserPhone: async (updateUserPhoneRequest: UpdateUserPhoneRequest) => {
      const { userApi } = getApi();
      const user = await userApi.updateCurrentUserPhone({ updateUserPhoneRequest });
      set({ user });
    },
  };
});
