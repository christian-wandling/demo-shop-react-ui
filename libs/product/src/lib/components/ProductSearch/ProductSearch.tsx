import { useNavigate } from 'react-router';
import { useProductStore } from '../../+state/useProductStore';

export const ProductSearch = () => {
  const navigate = useNavigate();
  const filter = useProductStore(state => state.filter);
  const { setFilter } = useProductStore();

  /**
   * Updates the product name filter in the application state
   *
   * @param name - The product name to filter by, or undefined to clear the filter
   */
  const setProductNameFilter = (name: string) => {
    setFilter({ name });
  };

  /**
   * Navigates the user to the products page
   *
   * This method is typically triggered when a user submits a search or clicks
   * on the search component.
   */
  const goToProductPage = () => {
    navigate('/products');
  };

  const handleKeyDown = (key: string) => {
    if (key === 'Enter') {
      goToProductPage();
    }
  };

  return (
    <div className="flex align-middle lg:ml-6 cursor-pointer relative">
      <a className="p-2.5 text-gray-400 hover:text-gray-500">
        <span className="sr-only">Search</span>
        <img src="/icons/search.svg" alt="search" height="20" width="20" />
      </a>
      <div className={`search-input relative ${filter.name ? 'extended' : ''}`}>
        <input
          type="text"
          name="price"
          className="block w-full my-0.5 rounded-md border-0 py-1.5 px-3 text-gray-900 ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm/6"
          placeholder="Search products"
          value={filter.name}
          onChange={e => setProductNameFilter(e.target.value)}
          onKeyDown={e => handleKeyDown(e.key)}
        />
        {filter.name && (
          <div className="enter-icon absolute right-2 inset-y-0 flex items-center" onClick={goToProductPage}>
            <img src="/icons/enter.svg" alt="enter" height="16" width="16" />
          </div>
        )}
      </div>
    </div>
  );
};
