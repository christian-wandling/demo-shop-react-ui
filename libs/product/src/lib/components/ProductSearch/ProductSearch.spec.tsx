import { render, screen, fireEvent } from '@testing-library/react';
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { ProductSearch } from './ProductSearch';
import { useProductStore } from '../../+state/useProductStore';
import { useNavigate } from 'react-router';

vi.mock('react-router', () => ({
  useNavigate: vi.fn().mockReturnValue(vi.fn()),
}));

const mockState = {
  filter: {},
  setFilter: vi.fn(),
};

vi.mock('../../+state/useProductStore', () => ({
  useProductStore: vi.fn(selectorOrState => {
    if (typeof selectorOrState === 'function') {
      return selectorOrState(mockState);
    }
    return mockState;
  }),
}));

describe('ProductSearch', () => {
  const store = useProductStore();
  const navigate = useNavigate();

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('should render successfully', () => {
    const { baseElement } = render(<ProductSearch />);
    expect(baseElement).toBeTruthy();
  });

  it('should match snapshot', () => {
    const { baseElement } = render(<ProductSearch />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should display the search icon', () => {
    render(<ProductSearch />);
    const searchIcon = screen.getByAltText('search');
    expect(searchIcon).toBeTruthy();
  });

  it('should render an input field with correct placeholder', () => {
    render(<ProductSearch />);
    const inputElement = screen.getByPlaceholderText('Search products');
    expect(inputElement).toBeTruthy();
  });

  it('should update filter when text is entered', () => {
    const setFilter = vi.spyOn(store, 'setFilter');

    render(<ProductSearch />);
    const inputElement = screen.getByPlaceholderText('Search products');

    fireEvent.change(inputElement, { target: { value: 'test product' } });

    expect(setFilter).toHaveBeenCalledWith({ name: 'test product' });
  });

  it('should navigate to products page when Enter key is pressed', () => {
    render(<ProductSearch />);
    const inputElement = screen.getByPlaceholderText('Search products');

    fireEvent.keyDown(inputElement, { key: 'Enter' });

    expect(navigate).toHaveBeenCalledWith('/products');
  });

  it('should not navigate when a key other than Enter is pressed', () => {
    render(<ProductSearch />);
    const inputElement = screen.getByPlaceholderText('Search products');

    fireEvent.keyDown(inputElement, { key: 'a' });

    expect(navigate).not.toHaveBeenCalled();
  });

  it('should display the enter icon when filter name is set', () => {
    mockState.filter = { name: 'test product' };

    render(<ProductSearch />);
    const enterIcon = screen.getByAltText('enter');
    expect(enterIcon).toBeTruthy();
  });

  it('should not display the enter icon when filter name is empty', () => {
    mockState.filter = { name: '' };

    render(<ProductSearch />);
    const enterIcon = screen.queryByAltText('enter');
    expect(enterIcon).not.toBeTruthy();
  });

  it('should navigate to products page when enter icon is clicked', () => {
    mockState.filter = { name: 'test product' };

    render(<ProductSearch />);
    const enterIcon = screen.getByAltText('enter');

    fireEvent.click(enterIcon);

    expect(navigate).toHaveBeenCalledWith('/products');
  });

  it('should apply "extended" class when filter name is set', () => {
    mockState.filter = { name: 'test product' };
    render(<ProductSearch />);
    const searchInput = document.querySelector('.search-input');
    expect(searchInput?.classList).toContain('extended');
  });

  it('should not apply "extended" class when filter name is empty', () => {
    mockState.filter = { name: '' };

    render(<ProductSearch />);
    const searchInput = document.querySelector('.search-input');
    expect(searchInput?.classList).not.toContain('extended');
  });
});
