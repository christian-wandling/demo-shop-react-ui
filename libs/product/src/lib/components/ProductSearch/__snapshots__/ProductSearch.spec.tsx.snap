// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductSearch > should match snapshot 1`] = `
<body>
  <div>
    <div
      class="search-container flex align-middle lg:ml-6 cursor-pointer relative"
    >
      <a
        class="p-2.5 text-gray-400 hover:text-gray-500"
      >
        <span
          class="sr-only"
        >
          Search
        </span>
        <img
          alt="search"
          height="20"
          src="/icons/search.svg"
          width="20"
        />
      </a>
      <div
        class="search-input relative "
      >
        <input
          class="block w-full my-0.5 rounded-md border-0 py-1.5 px-3 text-gray-900 ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm/6 transition-all duration-200"
          name="price"
          placeholder="Search products"
          type="text"
        />
      </div>
    </div>
  </div>
</body>
`;
